src/server/worker/pipeline/transform/registry_v2.go

```diff
@@ -0,0 +1,670 @@
+						pj.logger.Logf("error stack: %+v", st.StackTrace())
+					}
+					err = errors.Unwrap(err)
+				}
+				// Get job state, increment restarts, write job state
+				pj.ji, err = pj.driver.PachClient().InspectJob(pj.ji.Job.ID, false)
+				if err != nil {
+					return err
+				}
+				pj.ji.Restart++
+				if err := pj.writeJobInfo(); err != nil {
+					pj.logger.Logf("error incrementing restart count for job (%s): %v", pj.ji.Job.ID, err)
+				}
+				// Reload the job's commitInfo(s) as they may have changed and clear the state of the commit(s).
+				pj.commitInfo, err = reg.driver.PachClient().InspectCommit(pj.commitInfo.Commit.Repo.Name, pj.commitInfo.Commit.ID)
+				if err != nil {
+					return err
+				}
+				if err := pj.driver.PachClient().ClearCommitV2(pj.commitInfo.Commit.Repo.Name, pj.commitInfo.Commit.ID); err != nil {
+					return err
+				}
+				if metaCommit != nil {
+					pj.metaCommitInfo, err = reg.driver.PachClient().InspectCommit(metaCommit.Repo.Name, metaCommit.ID)
+					if err != nil {
+						return err
+					}
+					if err := pj.driver.PachClient().ClearCommitV2(pj.metaCommitInfo.Commit.Repo.Name, pj.metaCommitInfo.Commit.ID); err != nil {
+						return err
+					}
+				}
+				return nil
+			})
+			pj.logger.Logf("master done running processJobs")
+			// TODO: make sure that all paths close the commit correctly
+		}); err != nil {
+			return err
+		}
+		// This should block until the callback has completed
+		mutex.Lock()
+		return nil
+	})
+	go func() {
+		defer reg.limiter.Release()
+		if err := asyncEg.Wait(); err != nil {
+			pj.logger.Logf("fatal job error: %v", err)
+		}
+	}()
+	return nil
+}
+
+// superviseJob watches for the output commit closing and cancels the job, or
+// deletes it if the output commit is removed.
+func (reg *registryV2) superviseJob(pj *pendingJobV2) error {
+	defer pj.cancel()
+	ci, err := pj.driver.PachClient().PfsAPIClient.InspectCommit(pj.driver.PachClient().Ctx(),
+		&pfs.InspectCommitRequest{
+			Commit:     pj.ji.OutputCommit,
+			BlockState: pfs.CommitState_FINISHED,
+		})
+	if err != nil {
+		if pfsserver.IsCommitNotFoundErr(err) || pfsserver.IsCommitDeletedErr(err) {
+			// Stop the job and clean up any job state in the registry
+			if err := reg.killJob(pj, "output commit missing"); err != nil {
+				return err
+			}
+			// Output commit was deleted. Delete job as well
+			if _, err := pj.driver.NewSTM(func(stm col.STM) error {
+				// Delete the job if no other worker has deleted it yet
+				jobPtr := &pps.EtcdJobInfo{}
+				if err := pj.driver.Jobs().ReadWrite(stm).Get(pj.ji.Job.ID, jobPtr); err != nil {
+					return err
+				}
+				return pj.driver.DeleteJob(stm, jobPtr)
+			}); err != nil && !col.IsErrNotFound(err) {
+				return err
+			}
+			return nil
+		}
+		return err
+	}
+	if strings.Contains(ci.Description, pfs.EmptyStr) {
+		return reg.killJob(pj, "output commit closed")
+	}
+	return nil
+
+}
+
+func (reg *registryV2) processJob(pj *pendingJobV2) error {
```

Comment:
this might be a bit of overkill now that there's effectively one state that is valid for this function, it should probably be refactored out at some point